crates/astria-sequencer-relayer/src/finalization_pipeline/mod.rs

```diff
@@ -0,0 +1,194 @@
+use std::{
+    collections::HashMap,
+    mem,
+};
+
+use astria_sequencer_types::SequencerBlockData;
+use tendermint::Hash;
+
+pub(crate) mod head_candidate;
+pub(crate) mod pipeline_item;
+
+pub(crate) use head_candidate::HeadCandidate;
+pub(crate) use pipeline_item::PipelineItem;
+
+// head of the sequencer chain as observed on cometbft.
+#[derive(Default)]
+pub(crate) struct Head {
+    block: PipelineItem,
+}
+
+// pipeline can handle forks 1 height deep
+#[derive(Default)]
+pub(crate) struct FinalizationPipeline {
+    pub(crate) chain_head: Head,
+    pending: HashMap<Hash, PipelineItem>,
+    validator_finalized: Vec<SequencerBlockData>,
+}
+
+impl FinalizationPipeline {
+    pub(crate) fn submit(&mut self, new_block: HeadCandidate) {
+        let new_block: PipelineItem = new_block.into();
+
+        match new_block.parent_block_hash() {
+            Some(new_block_parent) => {
+                debug_assert!(new_block.height() > self.chain_head.block.height());
+
+                let steps = new_block.height() - self.chain_head.block.height();
+                if steps == 1 {
+                    debug_assert!(new_block_parent == self.chain_head.block.block_hash());
+
+                    self.pending.insert(new_block.block_hash(), new_block);
+                } else if steps == 2 {
+                    let pending_at_prev_height = mem::replace(
+                        &mut self.pending,
+                        HashMap::from([(new_block.block_hash(), new_block)]),
+                    );
+                    for competing_block in pending_at_prev_height.into_values() {
+                        debug_assert!(
+                            competing_block.height() == self.chain_head.block.height() + 1
+                        );
+
+                        if competing_block.block_hash() == new_block_parent {
+                            let old_head = mem::replace(
+                                &mut self.chain_head,
+                                competing_block.canonize().unwrap(),
```

Comment:
`unwrap()`? seems unsafe, pls either handle err or explain how it's infallible using `expect`